runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java

```diff
@@ -33,98 +46,175 @@
+            allUpdates.addAll(x.getWriteSet()
+                    .getOrDefault(streamID, Collections.emptyList()));
+        });
+
+        // Record that we have modified this proxy.
+        if (allUpdates.size() > 0) {
+            modifiedProxies.add(proxy);
+        }
+
+        // Apply all the updates from the optimistic version to the end
+        // of the list of all updates for this thread.
+        IntStream.range(object.getOptimisticVersionUnsafe(), allUpdates.size())
+                .mapToObj(allUpdates::get)
+                .forEachOrdered(wrapper -> {
+                    SMREntry entry  = wrapper.getEntry();
+                    Object res = proxy.getUnderlyingObject()
+                            .applyUpdateUnsafe(entry, true);
+                    wrapper.setUpcallResult(res);
+                    wrapper.setHaveUpcallResult(true);
+                });
     }
 
-    /**
-     * Buffer away an object update, adding it to the write set that will be generated
-     * in the resulting TXEntry.
+    /** Access the underlying state of the object.
      *
-     * @param proxy        The SMR Object proxy to buffer for.
-     * @param SMRMethod    The method being called.
-     * @param SMRArguments The arguments to that method.
-     * @param serializer   The serializer to use.
-     * @param <T>          The type of the proxy.
+     * @param proxy             The proxy making the state request.
+     * @param accessFunction    The access function to execute.
+     * @param <R>               The return type of the access function.
+     * @param <T>               The type of the proxy.
+     * @return                  The result of the access.
      */
-    @SuppressWarnings("unchecked")
     @Override
-    public <T> void bufferObjectUpdate(CorfuSMRObjectProxy<T> proxy, String SMRMethod,
-                                       Object[] SMRArguments, ISerializer serializer, boolean writeOnly) {
-        objectMap
-                .compute(proxy, (k, v) ->
+    public <R, T> R access(ICorfuSMRProxyInternal<T> proxy, ICorfuSMRAccess<R, T> accessFunction) {
+        // First, we add this access to the read set.
+        readSet.add(proxy.getStreamID());
+
+        // Next, we check if the write set has any
+        // outstanding modifications.
+        try {
+            return proxy.getUnderlyingObject().optimisticallyReadAndRetry((v, o) -> {
+                // to ensure snapshot isolation, we should only read from
+                // the first read timestamp.
+                if (v == getFirstReadTimestamp() &&
+                        objectIsOptimisticallyUpToDateUnsafe(proxy))
                 {
-                    TransactionalObjectData<T> data = v;
-                    if (v == null) {
-                        data = new TransactionalObjectData<>(proxy);
-                    }
+                    return accessFunction.access(o);
+                }
+                throw new ConcurrentModificationException();
```

Comment:
optimisticallyReadAndRetry expects a BiFunction with a return value. We could return null, but .access could legitimately return null. If the access was not possible (because the timestamp was wrong), we need to throw a concurrentmodificationexception to force a full write lock.